### DISCUSSION TITLE
Implementing `has_key` filter for sqlite JSON

### DISCUSSION THREAD

**Discussion by chrisjsewell** [sqlalchemy/sqlalchemy#7836] [Usage Questions]
Heya, I was looking to essentially implement https://docs.sqlalchemy.org/en/14/dialects/postgresql.html#sqlalchemy.dialects.postgresql.JSONB.Comparator.has_key for sqlite JSON.

The code below looks to work, using https://www.sqlite.org/json1.html#jeach and https://docs.sqlalchemy.org/en/14/tutorial/data_select.html#table-valued-functions, but it gives a warning.
Is this a problem?

```
SAWarning: SELECT statement has a cartesian product between FROM element(s) "anon_1" and FROM element "tbl".  Apply join condition(s) between each element to resolve.
```

```python
import sqlalchemy as sa

metadata = sa.MetaData()
my_table = sa.Table(
    "tbl",
    metadata,
    sa.Column('id', sa.Integer, primary_key=True),
    sa.Column('data', sa.JSON()),
)

engine = sa.create_engine('sqlite:///:memory:', future=True, echo=False)
metadata.create_all(engine)

with engine.begin() as conn:
    conn.execute(sa.insert(my_table).values(data={'a': None, 'c': 2, 'd': {'e': 1}}))
    conn.execute(sa.insert(my_table).values(data={'a': None, 'c': 2, 'd': {'f': 1, 'g': 2}}))

with engine.connect() as conn:
    sub_dict = my_table.c.data['d']
    has_key = sa.func.json_each(sub_dict).table_valued("key").c.key == 'f'
    result = conn.execute(sa.select(my_table.c.id).where(has_key)).all()
```

**Comment by chrisjsewell** [comment]
As an added bonus, I was also looking to implement https://docs.sqlalchemy.org/en/14/dialects/postgresql.html#sqlalchemy.dialects.postgresql.JSONB.Comparator.contains, but not sure where to start with his one?

**Reply by zzzeek** [reply]
that cartesian product means you want to JOIN from table valued has_key back to my_table.   im not deeply fluent with this kind of query but it looks odd that these two things arent being related together in some way.

**Reply by chrisjsewell** [reply]
Yeh, hmm, I see where you are coming from, but the https://www.sqlite.org/json1.html#jeach examples don't appear to require a join

**Reply by zzzeek** [reply]
the syntax is bizarre:

```
 SELECT tbl.id 
FROM tbl, json_each(JSON_QUOTE(JSON_EXTRACT(tbl.data, ?))) AS anon_1 
WHERE anon_1."key" = ?
```

I guess it is most similar to a JOIN LATERAL.

the only workaround I have right now is to join them as the docs state, SQLite hardly cares

```
    tv = sa.func.json_each(sub_dict).table_valued("key")
    has_key = tv.c.key == "f"
    result = conn.execute(sa.select(my_table.c.id).join(tv, sa.true()).where(has_key)).all()
```

**Reply by zzzeek** [reply]
I can just turn off the from linter for table-valued functions, will do that for next release

**Reply by chrisjsewell** [reply]
Cheers!

**Reply by zzzeek** [reply]
it needs to be a parameter.   only certain table valued functions have this property.   see #7845

**Reply by CaselIT** [reply]
> I can just turn off the from linter for table-valued functions, will do that for next release

I think pg doesn't like that. Also I guess it depends on the function. For example generate_series will cause a cartesian product unless joined. In general every function that returns more than a row has the same issue

**Reply by chrisjsewell** [reply]
The `joins_implicitly` construct looks sensible to me 👍 

(For my particular use case, the `.join(tv, sa.true())` approach would be a real pain, since my code constructs the "field selector" and "filters" independently)
